Version 9.x Reports Whitespace Before Paren After String with Emojis

I have read check documentation: https://checkstyle.org/config_whitespace.html#ParenPad
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```bash
/var/tmp $ javac Test.java

/var/tmp $ cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
  "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
  "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="ParenPad"/>
  </module>
</module>

/var/tmp $ cat Test.java
public class Test {
    void emojiFamilyRunner(Consumer<String> testRunner, String text) {
        testRunner.accept("👩‍👩‍👧‍👧 " + text); // count as 12 units
    }

    void emojiFamilyWithSkinToneModifierRunner(Consumer<String> testRunner, String text) {
        testRunner.accept("👩🏻‍👩🏽‍👧🏾‍👦🏿 " + text); // count as 20 units
    }
}

/var/tmp $ RUN_LOCALE="-Duser.language=en -Duser.country=US"
/var/tmp $ java $RUN_LOCALE -jar checkstyle-9.2-all.jar -c config.xml Test.java
Starting audit...
[ERROR] C:\Users\Work\Documents\Playground\CheckstylePlayground\.\src\main\java\Test.java:8:44: ')' is preceded with whitespace. [ParenPad]
[ERROR] C:\Users\Work\Documents\Playground\CheckstylePlayground\.\src\main\java\Test.java:12:48: ')' is preceded with whitespace. [ParenPad]
Audit done.
```

For Windows users, please use `type` instead of `cat` and run

```cmd
set RUN_LOCALE="-Duser.language=en -Duser.country=US"
java %RUN_LOCALE% -jar checkstyle-9.2-all.jar -c config.xml Test.java
```

in place of the last 2 commands above.

---

I would have expected no errors to be reported in this case but it appears the emoji messes up with where the paren is being seen. This doesn't occur when using 8.45.1.

This appears to be similar to https://github.com/checkstyle/checkstyle/issues/10920.

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/AbstractParenPadCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/AbstractParenPadCheck.java
@@ -86,14 +86,16 @@ public abstract class AbstractParenPadCheck
      */
     protected void processLeft(DetailAST ast) {
         final String line = getLines()[ast.getLineNo() - 1];
+        final int[] codePoints = line.codePoints().toArray();
         final int after = ast.getColumnNo() + 1;
-        if (after < line.length()) {
-            if (option == PadOption.NOSPACE
-                && Character.isWhitespace(line.charAt(after))) {
+
+        if (after < codePoints.length) {
+            final boolean hasWhitespaceAfter =
+                    CommonUtil.isCodePointWhitespace(codePoints, after);
+            if (option == PadOption.NOSPACE && hasWhitespaceAfter) {
                 log(ast, MSG_WS_FOLLOWED, OPEN_PARENTHESIS);
             }
-            else if (option == PadOption.SPACE
-                     && !Character.isWhitespace(line.charAt(after))
+            else if (option == PadOption.SPACE && !hasWhitespaceAfter
                      && line.charAt(after) != CLOSE_PARENTHESIS) {
                 log(ast, MSG_WS_NOT_FOLLOWED, OPEN_PARENTHESIS);
             }
@@ -109,13 +111,15 @@ public abstract class AbstractParenPadCheck
         final int before = ast.getColumnNo() - 1;
         if (before >= 0) {
             final String line = getLines()[ast.getLineNo() - 1];
-            if (option == PadOption.NOSPACE
-                && Character.isWhitespace(line.charAt(before))
+            final int[] codePoints = line.codePoints().toArray();
+            final boolean hasPrecedingWhitespace =
+                    CommonUtil.isCodePointWhitespace(codePoints, before);
+
+            if (option == PadOption.NOSPACE && hasPrecedingWhitespace
                 && !CommonUtil.hasWhitespaceBefore(before, line)) {
                 log(ast, MSG_WS_PRECEDED, CLOSE_PARENTHESIS);
             }
-            else if (option == PadOption.SPACE
-                && !Character.isWhitespace(line.charAt(before))
+            else if (option == PadOption.SPACE && !hasPrecedingWhitespace
                 && line.charAt(before) != OPEN_PARENTHESIS) {
                 log(ast, MSG_WS_NOT_PRECEDED, CLOSE_PARENTHESIS);
             }

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceBeforeCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceBeforeCheck.java
@@ -213,12 +213,12 @@ public class NoWhitespaceBeforeCheck
         final int before = ast.getColumnNo() - 1;
         final int[] codePoints = line.codePoints().toArray();
 
-        if ((before == -1 || isWhitespace(codePoints, before))
+        if ((before == -1 || CommonUtil.isCodePointWhitespace(codePoints, before))
                 && !isInEmptyForInitializerOrCondition(ast)) {
             boolean flag = !allowLineBreaks;
             // verify all characters before '.' are whitespace
             for (int i = 0; i <= before - 1; i++) {
-                if (!isWhitespace(codePoints, i)) {
+                if (!CommonUtil.isCodePointWhitespace(codePoints, i)) {
                     flag = true;
                     break;
                 }
@@ -227,23 +227,6 @@ public class NoWhitespaceBeforeCheck
                 log(ast, MSG_KEY, ast.getText());
             }
         }
-    }
-
-    /**
-     * Converts the Unicode code point at index {@code index} to it's UTF-16
-     * representation, then checks if the character is whitespace. Note that the given
-     * index {@code index} should correspond to the location of the character
-     * to check in the string, not in code points.
-     *
-     * @param codePoints the array of Unicode code points
-     * @param index the index of the character to check
-     * @return true if character at {@code index} is whitespace
-     */
-    private static boolean isWhitespace(int[] codePoints, int index) {
-        //  We only need to check the first member of a surrogate pair to verify that
-        //  it is not whitespace.
-        final char character = Character.toChars(codePoints[index])[0];
-        return Character.isWhitespace(character);
     }
 
     /**

--- a/src/main/java/com/puppycrawl/tools/checkstyle/utils/CommonUtil.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/utils/CommonUtil.java
@@ -626,4 +626,21 @@ public final class CommonUtil {
         return isInt;
     }
 
+    /**
+     * Converts the Unicode code point at index {@code index} to it's UTF-16
+     * representation, then checks if the character is whitespace. Note that the given
+     * index {@code index} should correspond to the location of the character
+     * to check in the string, not in code points.
+     *
+     * @param codePoints the array of Unicode code points
+     * @param index the index of the character to check
+     * @return true if character at {@code index} is whitespace
+     */
+    public static boolean isCodePointWhitespace(int[] codePoints, int index) {
+        //  We only need to check the first member of a surrogate pair to verify that
+        //  it is not whitespace.
+        final char character = Character.toChars(codePoints[index])[0];
+        return Character.isWhitespace(character);
+    }
+
 }

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/whitespace/ParenPadCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/whitespace/ParenPadCheckTest.java
@@ -492,6 +492,21 @@ public class ParenPadCheckTest
                 getNonCompilablePath("InputParenPadCheckRecordsSpace.java"), expected);
     }
 
+    @Test
+    public void testParenPadCheckEmoji() throws Exception {
+
+        final String[] expected = {
+            "25:45: " + getCheckMessage(MSG_WS_PRECEDED, ")"),
+            "29:49: " + getCheckMessage(MSG_WS_PRECEDED, ")"),
+            "33:26: " + getCheckMessage(MSG_WS_FOLLOWED, "("),
+            "37:26: " + getCheckMessage(MSG_WS_FOLLOWED, "("),
+            "43:9: " + getCheckMessage(MSG_WS_FOLLOWED, "("),
+            "43:61: " + getCheckMessage(MSG_WS_PRECEDED, ")"),
+        };
+        verifyWithInlineConfigParser(
+                getPath("InputParenPadCheckEmoji.java"), expected);
+    }
+
     /**
      * Pitest requires us to specify more concrete lower bound for condition for
      * ParenPadCheck#isAcceptableToken as nodes of first several types like CTOR_DEF,

--- a/src/test/java/com/puppycrawl/tools/checkstyle/utils/CommonUtilTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/utils/CommonUtilTest.java
@@ -19,6 +19,7 @@
 
 package com.puppycrawl.tools.checkstyle.utils;
 
+import static com.google.common.truth.Truth.assertThat;
 import static com.puppycrawl.tools.checkstyle.internal.utils.TestUtil.isUtilsClassHasPrivateConstructor;
 import static org.hamcrest.CoreMatchers.containsString;
 import static org.hamcrest.CoreMatchers.is;
@@ -487,6 +488,15 @@ public class CommonUtilTest extends AbstractPathTestSupport {
         final Configuration config = ConfigurationLoader.loadConfiguration(uri.toString(),
             new PropertiesExpander(properties));
         assertEquals("Checker", config.getName(), "Unexpected config name!");
+    }
+
+    @Test
+    public void testIsCodePointWhitespace() {
+        final int[] codePoints = " 123".codePoints().toArray();
+        assertThat(CommonUtil.isCodePointWhitespace(codePoints, 0))
+                .isTrue();
+        assertThat(CommonUtil.isCodePointWhitespace(codePoints, 1))
+                .isFalse();
     }
 
     private static class TestCloseable implements Closeable {

--- a/src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/parenpad/InputParenPadCheckEmoji.java
+++ b/src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/parenpad/InputParenPadCheckEmoji.java
@@ -1,0 +1,45 @@
+/*
+ParenPad
+option = (default)nospace
+tokens = (default)ANNOTATION, ANNOTATION_FIELD_DEF, CTOR_CALL, CTOR_DEF, DOT, \
+         ENUM_CONSTANT_DEF, EXPR, LITERAL_CATCH, LITERAL_DO, LITERAL_FOR, LITERAL_IF, \
+         LITERAL_NEW, LITERAL_SWITCH, LITERAL_SYNCHRONIZED, LITERAL_WHILE, METHOD_CALL, \
+         METHOD_DEF, QUESTION, RESOURCE_SPECIFICATION, SUPER_CTOR_CALL, LAMBDA, RECORD_DEF
+
+
+*/
+package com.puppycrawl.tools.checkstyle.checks.whitespace.parenpad;
+
+import java.util.function.Consumer;
+
+public class InputParenPadCheckEmoji {
+    void emojiFamilyRunner1(Consumer<String> testRunner, String text) {
+        testRunner.accept("👩‍👩‍👧‍👧 " + text); // ok
+    }
+
+    void emojiFamilyWithSkinToneModifierRunner1(Consumer<String> testRunner, String text) {
+        testRunner.accept("👩🏻‍👩🏽‍👧🏾‍👦🏿 " + text); // ok
+    }
+
+    void emojiFamilyRunner2(Consumer<String> testRunner, String text) {
+        testRunner.accept("👩‍👩‍👧‍👧 " + text ); // violation '')' is preceded with whitespace'
+    }
+
+    void emojiFamilyWithSkinToneModifierRunner2(Consumer<String> testRunner, String text) {
+        testRunner.accept("👩🏻‍👩🏽‍👧🏾‍👦🏿 " + text ); // violation '')' is preceded with whitespace'
+    }
+
+    void emojiFamilyRunner3(Consumer<String> testRunner, String text) {
+        testRunner.accept( "👩‍👩‍👧‍👧 " + text); // violation ''(' is followed by whitespace'
+    }
+
+    void emojiFamilyWithSkinToneModifierRunner3(Consumer<String> testRunner, String text) {
+        testRunner.accept( "👩🏻‍👩🏽‍👧🏾‍👦🏿 " + text); // violation ''(' is followed by whitespace'
+    }
+
+    void emojiFamilyWithSkinToneModifierRunnerTricky(Consumer<String> testRunner, String text) {
+        testRunner
+        .accept
+        ( "👩🏻‍👩🏽‍👧🏾‍👦🏿 "  + "ab cdefg" + "👩🏽‍👧🏾‍👦🏿 " + text ); // 2 violations
+    }
+}

